Allow installation in combination with PHPUnit < 9.1

jrfnl opened the pull request:
### Add custom autoloader/fall-through to PHPUnit native functionality

This commit:
* Adds a custom autoloader which will either load the ProphecyTrait or will load an empty trait which will fall through to using the PHPUnit native functionality for PHPUnit 4.x - 9.0.
* Widens the version requirements for PHP and PHPUnit to allow installing of this package on PHP 5.4 - current in combination with PHPUnit 4.8.36/5.7.21 - current.
* Updates the `autoload` section in the `composer.json` to load the custom autoloader instead of using PSR4 autoloading.
* Updates the README with information about the fact that the package no longer restricts installation to PHPUnit 9.

### Add availability tests for the fall-through functionality

This commit:
* Adds an additional test class which make sure that the `ProphecyTrait` functionality can be safely used and falls through to the PHPUnit native functionality for PHPUnit < 9.1.
* Prevents the _real_ unit tests from loading on PHP < 7.3 via an annotation in the `phpunit.xml.dist` file.
    If those files would be loaded on older PHP versions, the test run would break on the parse errors due to modern PHP syntax being used.
* Prevents the _real_ unit tests from running on PHPUnit < 9.1.
    These tests only need to be run on PHPUnit >= 9.1 as that is the only time when the polyfills will actually be used.

### GH Actions: adjust the test workflow

... to:
* Run the availability tests against all PHP versions on which the package can be installed with `prefer-stable` (`highest`).
* Run the availability tests, as well as the _real_ tests against `lowest` and `stable` for PHPUnit 9.1+ on PHP 7.3 - 8.1.

The test selection is done automatically via the `phpunit.xml.dist` file as adjusted in the previous commit.

As for the running of the _real_ tests against the `lowest` supported PHPUnit 9 version, this requires some tweaking to the workflow script.
As PHPUnit didn't always have an explicit requirement for PHP, running with `prefer-lowest` on PHP 7.3 - 8.1 with the current PHPUnit version constraints would install PHPUnit 4.8.
So, for the `prefer-lowest` runs, we need to make sure that `lowest` is correctly interpreted as "lowest" within the PHPUnit 9.1+ range.

### ~~[Optional] Tests: use custom bootstrap file~~ (Removed)

stof commented:
Rather than defining a custom autoloader, can't we use a conditional class definition in the ProphecyTrait.php file itself, which keeps things compatible with the Composer autoloader (and also with other autoloaders than might be used, for instance if someone builds a phar) ?

for the bootstrap file, please keep it out of that PR (btw, I'm not convinced about the need for such `PATH_TO_PROPHECY` variable. There are already ways to use a local clone of prophecy, either by configurating a composer path repository or just by adding replacing the folder inside `vendor` by a symlink).

jrfnl commented:
> Rather than defining a custom autoloader, can't we use a conditional class definition in the ProphecyTrait.php file itself, which keeps things compatible with the Composer autoloader (and also with other autoloaders than might be used, for instance if someone builds a phar) ?

Unfortunately, no.

For this PR, I've used the same PHP and PHPUnit version requirements as for the PHPUnit Polyfills package (and the `rdohms/phpunit-arraysubset-asserts` package for that matter): `"php": "^5.4 || ^7.0 || ^8.0"` and `"phpunit/phpunit": "^4.8.36 || ^5.7.21 || ^6.0 || ^7.0 || ^8.0 || ^9.1"`.
This allows for these packages to be used as companion add-ons.

The (real, non-empty) `ProphecyTrait.php` file contains nullable types and return types, which means loading that file would result in parse errors on PHP < 7.1, even when the real/empty trait would be created conditionally.

It would still allow to widen the requirement, but in a far more limited manner (`"php": "^7.1 || ^8.0"` and `"phpunit/phpunit": "^7.0 || ^8.0 || ^9.1"`) than currently implemented.

As for removing those type declarations: that would make the Trait incompatible with PHPUnit 9.x and only allow for it to be used with PHPUnit 10.x in which the PHPUnit native `prophesize()` method has been removed, leaving people on PHPUnit 9.x to still receive the deprecation warnings.


> for the bootstrap file, please keep it out of that PR (btw, I'm not convinced about the need for such `PATH_TO_PROPHECY` variable. There are already ways to use a local clone of prophecy, either by configurating a composer path repository or just by adding replacing the folder inside `vendor` by a symlink).

No problem, I put it in a separate commit anyway for that reason, so I will just remove that commit from the PR.

jrfnl commented:
@stof Just checking: did my answer above clarify the "why this way?" of the setup of this PR sufficiently ?

stof commented on .github/workflows/ci.yml:
```diff
@@ -49,10 +60,16 @@ jobs:
 
+      - name: 'Composer: update PHPUnit for testing lowest'
+        if: ${{ matrix.dependency-versions == 'lowest' }}
+        run: composer require --no-update phpunit/phpunit:"^9.1"
```
changing the requirement before running with lowest deps defeats the goal of this job, which is to ensure that the lowest bound of the requirements is the right one.

We might instead add a conflict rule for `phpunit/phpunit <9.1`, which still keeps the requirement untouched.

jrfnl replied:
Wouldn't a conflict rule prevent installation in combination with PHPUnit < 9.1 ?

stof replied:
a conflict rule being added here for this job (instead of replacing the `require` rule as you do), not a rule added in the committed composer.json

jrfnl replied:
Ah! In that case, my question changes: how can I set a `conflict` rule from the command-line ? I'm not aware of any Composer CLI command available to do this. Interested to learn this new trick ;-)

stof replied:
I don't think the CLI allows to do it. You will need to do some sed magic on the composer.json file.

jrfnl replied:
Hmm.. that sounds like a very complicated way to achieve the same thing as is being achieved via the above command...

I suppose I could do it by adding a second `test-lowest-composer.json` file to the repo and conditionally using that during the CI runs, but that would risk that file getting out-of-sync with the _real_ `composer.json` files.

Other than that, I'm out of ideas. I don't have the set up to create or test `sed` commands, so getting that working will be _challenging_.

stof replied:
The current setup does not catch mistakes in the lowest bound in the composer.json, which is the goal of a `--prefer-lowest` CI job. Using a separate composer.json would not help either unless you have a test enforcing it to stay in sync with the normal composer.json.
As we don't have a `--prefer-lowest` job in our current CI setup, another solution is to avoid adding such a job in your PR (as that's not its main goal)

jrfnl replied:
> The current setup does not catch mistakes in the lowest bound in the composer.json, which is the goal of a --prefer-lowest CI job. 

Actually, no. My _intention_ with adding the `lowest` builds was to ensure that the _real_ trait works as advertised in the high/low supported versions of PHPUnit. Nothing more.

But yeah, I can remove it from this PR for now and you can revisit this later.

stof commented on prophecy-phpunit-autoload.php:
```diff
@@ -0,0 +1,64 @@
+     *
+     * {@internal The code in this file must be PHP cross-version compatible for PHP 5.4 - current!}
+     */
+    class Autoload
```
This whole autoload stuff could be replaced by having a `ProphecyTrait.php` file that conditionally requires the right file based on the PHPUnit version (and a fake trait definition in a `if (false)` in that file, to be compatible with the optimized composer autoloader) instead of having a file being loaded eagerly to register the autoloader.

In the case of the Yoast polyfills, I understand the usage of an autoloader given the amount of traits they have. But for a single trait, this is overkill.

jrfnl replied:
@stof I hear what you are saying.

If I understand you correctly, you'd like the `files` directive to load the `ProphecyTrait.php`file, which will then contain just an if/else to either load the real trait or declare an empty trait based on the detected PHPUnit version ?

The downside of that is that the (empty) trait will then always be declared as soon as the `vendor/autoload.php` file is included, even when not in a unit test situation.

I suppose I could make the declaration of the empty trait conditional on the PHPUnit version method even being available, but I somehow don't think that's what you meant ?

stof replied:
I'd like to keep relying on the Composer autoloader instead of using the `files` section

jrfnl replied:
Any preference for the file name of the _real_ trait (as it will have to be renamed) ? Would `ProphecyTraitReal.php` work for you ?

jrfnl replied:
Also, as a side-note, which I presume you already considered:
The trade-off of removing the autoloader is that _if_ at some point in the future for PHPUnit 10 or PHP 9, you'd need to introduce a slightly different trait again, the `ProphecyTrait.php` file will just get messier with all the different conditions, but hey, each to their own. I respect your preference as the maintainer of this repo and will make the necessary adjustments.

stof replied:
Well, if we need that in the future and we use an autoloader, the same extra logic will still be needed in the autoloader.

jrfnl replied:
Just to verify: this would mean changing the `autoload` from `psr-4` (old)/`files` (current) to `classmap` to load the `ProphecyTrait.php` file when needed. Did I interpret that correctly ?

stof replied:
well, the `ProphecyTrait.php` file could still be loaded using psr-4 rather than classmap.

and for the 2 non-psr-4 files defining the actual implementations, I suggest naming them `ProphecyTrait.empty.php` and `ProphecyTrait.actual.php`. Using a `.` delimiter rather than a suffix in the file name will ensure that those files are excluded from any PSR-4 loading trying to make them loaded for `Prophecy\PhpUnit\ProphecyTraitEmpty`

jrfnl replied:
I've implemented this now as discussed above. It does feel a bit hacky for "abusing" PSR-4 essentially to do the autoloading, but heck, it seems to work.

stof commented on tests/Availability/AvailabilityTest.php:
```diff
@@ -0,0 +1,29 @@
+ *
+ * @coversNothing
+ */
+final class AvailabilityTest extends TestCase
```
this tes is not implemented as a proper testcase for the availability. It won't produce a testsuite failure if the trait is not available, but a fatal error in the test loading.

The availability test should look like that:

```php
final class AvailabilityTest extends TestCase
{
    public function testTraitIsAvailable()
    {
        $this->assertTrue(trait_exists(ProphecyTrait::class);
    }
}
```

jrfnl replied:
I implemented it as an availability test for the `prophesize()` _method_, not for the _trait_, though that is indirectly also tested as the trait is _used_ by this test, so if it would not be available, the test would error out on a fatal.

Maybe I should adjust the test description in the docblock to make that clearer?

jrfnl replied:
An alternative implementation of the test would IMO look like this:
```php
final class AvailabilityTest extends TestCase
{
    use ProphecyTrait;

    public function testTraitIsAvailable()
    {
        $this->assertTrue(method_exists($this, 'prophesize'));
    }
}
```

stof replied:
this is still not the right thing. You cannot use the current test case as the subject being tested. This does not play well with failure reporting.

The case of checking that the `prophesize` method exist means that we should indeed have a second assertion using a fixture class (as done in the `ProphecyTraitTest`) to check for the method existence.
but to me, the `trait_exists` existence check (*before* using the fixture class using that trait) still makes sense, to have a proper failure if the trait does not exist properly (due to a mistake in loading it) rather than a fatal error.

stof commented on .github/workflows/ci.yml:
```diff
@@ -34,9 +34,20 @@ jobs:
+          - php: '7.4'
+            dependency-versions: 'lowest'
+          - php: '8.0'
+            dependency-versions: 'lowest'
```
No need to have lowest jobs for each PHP version. We can simplify the matrix.

jrfnl replied:
Not sure I understand. Or do you mean to say, you don't want to test against `lowest` ?

The `lowest` runs are only relevant to test the actual polyfill trait in combination with PHPUnit 9.1+ and the PHP versions supported by that.
Running against `lowest` for the availabiity test would end up testing the PHPUnit native implementation across versions, which IMO is not the task of this project.

stof replied:
We don't need to have 4 jobs running against `lowest`. One of them is enough.

stof commented on prophecy-phpunit-autoload.php:
```diff
@@ -0,0 +1,64 @@
+
+namespace Prophecy\PhpUnit;
+
+use PHPUnit\Runner\Version as PHPUnit_Version;
```
no need to alias it.

jrfnl commented:
Thanks for the review @stof ! I've left some questions/responses to verify my understanding of your feedback.

jrfnl commented:
@stof I believe I've addressed all the feedback with the new commits. Hope this complies with how you want it.
Let me know if you want me to squash the new commits into the original ones for a cleaner history.

jrfnl commented:
Oh and one side-note: while the availability test does run and pass on PHPUnit 4.x, it is marked as risky now:

> 1) Prophecy\PhpUnit\Tests\Availability\AvailabilityTest::testSuccessfullyCallingProphesizeMethod
>Test code or tested code did not (only) close its own output buffers

jrfnl commented:
(Repushed to get rid of some tabs which had sneaked in. Naughty tabs!)

stof commented:
The test probably needs to define the `PHPUNIT_TESTSUITE` constant, as it runs a testcase as subject under test. I think that will remove that warning.

stof commented on src/ProphecyTrait.php:
```diff
@@ -1,110 +1,41 @@
+ *
+ * @return string Version number as a string.
+ */
+function getPHPUnitVersion()
```
should be tagged as `@internal` to exclude it from the BC policy.

stof commented on src/ProphecyTrait.php:
```diff
@@ -1,110 +1,41 @@
 
-        return $this->prophet;
-    }
+if (\version_compare(namespace\getPHPUnitVersion(), '9.1.0', '>=')) {
```
you can use `getPHPUnitVersion()` without the `namespace\`, as it is in the current namespace.

stof commented on tests/Availability/AvailabilityTest.php:
```diff
@@ -0,0 +1,31 @@
+{
+    public function testSuccessfullyCallingProphesizeMethod()
+    {
+        $this->assertTrue(trait_exists('Prophecy\PhpUnit\ProphecyTrait'), 'Failed to assert that the ProphecyTrait is avialable');
```
```suggestion
        $this->assertTrue(trait_exists('Prophecy\PhpUnit\ProphecyTrait'), 'Failed to assert that the ProphecyTrait is available');
```

jrfnl commented:
> 
> 
> The test probably needs to define the `PHPUNIT_TESTSUITE` constant, as it runs a testcase as subject under test. I think that will remove that warning.

It did indeed 🎉 . Note: I've used the `@before` annotation with an arbitrary method name to prevent having to add the `void` return type.

jrfnl commented:
@stof Anything I can do to help move this forward ?

jrfnl commented:
@stof Just checking if there is anything on my side this PR is waiting for ?

jrfnl commented:
Just checking in again: anything I can do to move this PR forward ?

jrfnl commented:
@stof What is this waiting for... ?

jrfnl commented:
Only 6 more days and then this PR has a birthday.... is there any hope of getting this merged and shall I just close this and call it a (wasted) day ?

aik099 commented:
> Only 6 more days and then this PR has a birthday.... is there any hope of getting this merged and shall I just close this and call it a (wasted) day ?

@jrfnl I've approved this PR, but I don't have any permission to merge it.

It's sad to admit, that nowadays people tend to use all the latest PHP and PHPUnit versions, and creating a library, that can work across older PHP/PHPUnit versions isn't as popular. People doing so are considered outcasts or worse.

In any case, let's wait a little longer because it costs nothing to keep an open PR, which someday might be even merged.

jrfnl commented:
> It's sad to admit, that nowadays people tend to use all the latest PHP and PHPUnit versions, and creating a library, that can work across older PHP/PHPUnit versions isn't as popular.

I hear you and I fully encourage that, but for those who are trying to upgrade to get there, it is still very helpful if they can least run their tests on both old vs new.

In the end, this PR doesn't actually enable the functionality of this repo for old PHP/PHPUnit versions, it just allows for it to be installed in combination with old PHP/PHPUnit versions, which makes those kind of migrations a lot easier.

jrfnl commented:
This PR will go to kindergarden soon... Enough.

@stof Next time you state that you would [welcome a PR](https://github.com/Yoast/PHPUnit-Polyfills/issues/62#issuecomment-957336131) for something, it might be nice to actually act on it. Now you've just wasted my time.
